### PR TITLE
GH-1806: As a Xtext committer I want to review, modify and approve changes regarding the LSP builder

### LINKED ISSUE #1806: As a smith I want the LSP builder changes to be reviewed and modified from an Xtext committers viewpoint
Note that there are a few TODO's in the code for this user story (search for IDE-3402).

### PR DESCRIPTION
**Pull request by szarnekow**
closes #1806

### CODE REVIEW

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/BuilderFrontend.java [hunk 1]
```diff
@@ -95,9 +117,11 @@ protected Void doInitialBuild() {
 		} finally {
 			LOG.info("Initial build done after " + sw);
 		}
-		return null;
 	}
 
+	/**
+	 * Trigger a clean build in the background.
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/BuilderFrontend.java:123]: we noted: a "clean build" cleans only

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/XWorkspaceManager.java [hunk 2]
```diff
@@ -114,11 +114,48 @@ synchronized protected void setWorkspaceConfig(XIWorkspaceConfig workspaceConfig
...
+	public Set<? extends XIProjectConfig> getProjectConfigs() {
+		XIWorkspaceConfig config = getWorkspaceConfig();
+		if (config == null) {
+			return Collections.emptySet();
+		}
+		return config.getProjects();
+	}
+
+	/**
+	 * Updates the workspace according the the updated information in the file with the given URI.
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/XWorkspaceManager.java:134]: typo: `the the`

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/XWorkspaceManager.java [hunk 3]
```diff
@@ -114,11 +114,48 @@ synchronized protected void setWorkspaceConfig(XIWorkspaceConfig workspaceConfig
...
+		}
+		return config.update(changedFile,
+				projectName -> getProjectBuilder(projectName).getProjectDescription());
+	}
+
+	/**
+	 * Answers true, if the uri is from a source folder.
+	 */
+	public boolean isSourceFile(URI uri) {
+		IProjectConfig projectConfig = getWorkspaceConfig().findProjectContaining(uri);
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/XWorkspaceManager.java:149]: note that below in `getProjectConfig(URI)` a null check is performed on the result of `getWorkspaceConfig()`

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/XURIBasedFileSystemAccessFactory.java [hunk 4]
```diff
@@ -31,7 +31,7 @@
 /** Creates an {@link IFileSystemAccess file system access} that is backed by a {@link URIConverter}. */
 @SuppressWarnings("restriction")
 @Singleton
-public class XURIBasedFileSystemAccessFactory {
+class XURIBasedFileSystemAccessFactory {
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/XURIBasedFileSystemAccessFactory.java:34]: Isn't package accessibility a problem for people who want to customize Xtext?
If not, could we reduce accessibility to "package" for other classes in ...xtext.server.build as well?

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/xtext/workspace/ProjectConfigSnapshot.java [hunk 5]
```diff
@@ -10,57 +10,64 @@
...
 	public ImmutableSet<String> getDependencies() {
 		return dependencies;
 	}
 
-	@Override
-	public ImmutableCollection<? extends ISourceFolderSnapshot> getSourceFolders() {
+	/**
+	 * Return the project source folders.
+	 */
+	public ImmutableSet<? extends SourceFolderSnapshot> getSourceFolders() {
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/xtext/workspace/ProjectConfigSnapshot.java:69]: Why the change from collection to set? Since, SourceFolderSnapshot does not implement `#equals()`, a set does not provide any benefit (and might actually be misleading).

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/xtext/workspace/WorkspaceConfigSnapshot.java [hunk 6]
```diff
@@ -40,38 +44,44 @@ public WorkspaceConfigSnapshot(XIWorkspaceConfig workspace) {
...
 	public URI getPath() {
 		return path;
 	}
 
-	@Override
-	public ImmutableCollection<? extends IProjectConfigSnapshot> getProjects() {
+	/**
+	 * Get all the projects known in this snapshot.
+	 */
+	public ImmutableSet<? extends ProjectConfigSnapshot> getProjects() {
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/xtext/workspace/WorkspaceConfigSnapshot.java:74]: same as above (also for Map vs. BiMap)

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/ResourceTaskManager.java [hunk 7]
```diff
@@ -266,7 +286,10 @@ protected ResourceTaskContext doCreateContext(URI uri, boolean isTemporary) {
 
 	/** Internal removal of all information related to a particular resource task context. */
 	protected synchronized void discardContextInfo(URI uri) {
-		uri2RTCs.remove(uri);
+		ResourceTaskContext result = uri2RTCs.remove(uri);
+		if (result != null) {
+			result.alive = false;
```
[the remark this answers is not in the document]
Comment (szarnekow) [reply plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/ResourceTaskManager.java]: Reworded to 
```
It is discouraged to complete the future normally if cancellation was detected. In other words: If the task produces a list of result items and detects cancellation requests half-way through the computation, it is discouraged to return the incomplete list. Instead an CancellationException or an OperationCanceledException should be thrown and propagated.
```

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/QueuedExecutorService.java [hunk 8]
```diff
@@ -153,6 +156,8 @@ protected QueuedTaskFuture(QueuedTask<T> task) {
 		 * marked as cancelled, but it lies in the discretion of the task's implementation how to react to that. This
 		 * future might even still complete normally if the task implementation chooses to ignore the cancellation
 		 * status of the cancel indicator.
+		 *
+		 * It is discouraged to complete semi-normally if cancellation was detected.
```
[the remark this answers is not in the document]
Comment (szarnekow) [reply plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/QueuedExecutorService.java:160]: Reworded to 
```
It is discouraged to complete the future normally if cancellation was detected. In other words: If the task produces a list of result items and detects cancellation requests half-way through the computation, it is discouraged to return the incomplete list. Instead an CancellationException or an OperationCanceledException should be thrown and propagated.
```

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/BuilderFrontend.java (hunk 1 above)
Comment (szarnekow) [reply plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/BuilderFrontend.java:123]: Clarified in the comment.

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/XWorkspaceManager.java (hunk 2 above)
Comment (szarnekow) [reply plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/XWorkspaceManager.java:134]: Fixed

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/XWorkspaceManager.java (hunk 3 above)
Comment (szarnekow) [reply plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/XWorkspaceManager.java:149]: :+1: Added null-guard.

File: plugins/org.eclipse.n4js.cli/src/org/eclipse/n4js/cli/compiler/N4jscCompiler.java [hunk 9]
```diff
@@ -141,15 +146,15 @@ private void setupWorkspaceBuildActionListener() {
 	}
 
 	private void warnIfNoProjectsFound() {
-		Set<? extends IProjectConfig> projects = workspaceManager.getWorkspaceConfig().getProjects();
+		Set<? extends IProjectConfig> projects = workspaceManager.getProjectConfigs();
```
[the remark this answers is not in the document]
Comment (szarnekow) [reply plugins/org.eclipse.n4js.cli/src/org/eclipse/n4js/cli/compiler/N4jscCompiler.java:149]: Yes, we should. Indeed the injection / lifecycle / dependencies of the N4jscCompiler could need some cleanup.

[the remark this answers is not in the document]
Comment (szarnekow) [reply plugins/org.eclipse.n4js.cli/src/org/eclipse/n4js/cli/compiler/N4jscCompiler.java:149]: Filed follow up #1815

File: plugins/org.eclipse.n4js.cli/src/org/eclipse/n4js/cli/lsp/LspServer.java [hunk 10]
```diff
@@ -160,7 +160,6 @@ private void run(XLanguageServerImpl languageServer, Builder<LanguageClient> lsB
 		Futures.getUnchecked(future);
 
 		N4jscConsole.println("Shutdown connection to LSP client");
-		languageServer.getQueuedExecutorService().shutdown();
```
[the remark this answers is not in the document]
Comment (szarnekow) [reply plugins/org.eclipse.n4js.cli/src/org/eclipse/n4js/cli/lsp/LspServer.java:163]: Reworded to `Connection to LSP client was shutdown`

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/LanguageServerFrontend.java [hunk 11]
```diff
@@ -92,11 +92,45 @@
...
+	public CompletableFuture<? extends Object> shutdown() {
+		return builderFrontend.shutdown();
+	}
+
+	/** Connect this front-end to the given client. */
+	public void connect(LanguageClient client) {
+		textDocumentFrontend.connect(client);
+	}
+
+	/** Disconnect this front-end from the currently connected. */
```
[the remark this answers is not in the document]
Comment (szarnekow) [reply plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/LanguageServerFrontend.java:130]: ;)

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/QueuedExecutorService.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (szarnekow) [reply plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/QueuedExecutorService.java:160]: Was reworded to

> It is discouraged to complete the future normally if cancellation was detected. In other words: If the task produces a list of result items and detects cancellation requests half-way through the computation, it is discouraged to return the incomplete list. Instead an CancellationException or an OperationCanceledException should be thrown and propagated.

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/QueuedExecutorService.java [hunk 12]
```diff
@@ -225,15 +230,15 @@ protected synchronized void doSubmitAllPending() {
...
+		if (submittedTasks.putIfAbsent(task.queueId, task) != null) {
 			throw new IllegalStateException("executor inconsistency: queue ID already in progress: " + task.queueId);
 		}
 		delegate.submit(task); // will eventually invoke #onDone()
 	}
 
 	/** Invoked by each running task upon completion, see {@link QueuedTask#run()}. */
 	protected synchronized void onDone(QueuedTask<?> task) {
-		if (runningTasks.remove(task.queueId) == null) {
+		if (submittedTasks.remove(task.queueId) != task) {
```
[the remark this answers is not in the document]
Comment (szarnekow) [reply plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/QueuedExecutorService.java:241]: Done

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/QueuedExecutorService.java [hunk 13]
```diff
@@ -334,7 +332,7 @@ public synchronized String stringify() {
 			}
 		}
 		sb.append(QueuedExecutorService.class.getSimpleName() + " showing all " + QueuedTask.class.getSimpleName());
-		sb.append("\nActive Tasks Queues (First task is running, succeeding tasks are waiting):\n");
+		sb.append("\nActive Tasks Queues (First task is submitted, succeeding tasks are waiting):\n");
```
[the remark this answers is not in the document]
Comment (szarnekow) [reply plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/QueuedExecutorService.java:335]: Done

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/ResourceTaskManager.java [hunk 14]
```diff
@@ -68,6 +75,11 @@
 	 */
 	protected final ThreadLocal<ResourceTaskContext> currentContext = new ThreadLocal<>();
 
+	/*
+	 * Review feedback:
+	 *
+	 * This looks like being the same as the the ConcurrentIndex structure?
```
[the remark this answers is not in the document]
Comment (szarnekow) [reply plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/ResourceTaskManager.java:81]: Understood. I'm afraid this will undergo some bigger restructuring in subsequent pull requests. Therefore I left the note here.

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/ResourceTaskContext.java [hunk 15]
```diff
@@ -62,6 +60,22 @@
...
+	 * updating resources after changes happened. We are implementing a different incremental build semantics in
+	 * refreshContext with fewer optimizations and therefore a performance drag.
+	 *
+	 * The logic, how a given resource set is updated after an incoming change is recorded, should certainly be
+	 * implemented exactly once.
+	 *
+	 * For xtext.ui, we made the mistake of duplicating this between the Eclipse builder and the reconciler, and this
+	 * was a constant PITA.
+	 *
+	 * #prepareRefresh is sort-of a brute force approach that has potential for optimizations.
```
[the remark this answers is not in the document]
Comment (szarnekow) [reply plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/ResourceTaskContext.java:76]: Fair point indeed. The Eclipse builder can simply "make a mess" with its current resource set in any case, since it will be discarded after the build was done. The drawback of this is the high setup cost for each build run. In Eclipse, we rely on the fact that the AutoBuildJob will only be triggered with certain delay which mitigates this cost-issue somehow.

Measurements in other language context showed though, that we have extremely short built times in LSP due to the reused resource set. That is something we should really try to keep as a non-functional property.

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/ResourceTaskManager.java [hunk 16]
```diff
@@ -47,8 +53,9 @@
  * Manages a set of {@link ResourceTaskContext}s, including creation, disposal, and executing tasks within those
  * contexts.
  */
+@SuppressWarnings("restriction")
 @Singleton
-public class ResourceTaskManager {
+public class ResourceTaskManager implements IReferenceFinder.IResourceAccess {
```
[the remark this answers is not in the document]
Comment (szarnekow) [reply plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/ResourceTaskManager.java:58]: Agreed. I reverted that change.

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/ResourceTaskManager.java [hunk 17]
```diff
@@ -78,11 +90,16 @@
...
-	protected IWorkspaceConfigSnapshot workspaceConfig = null;
+	protected WorkspaceConfigSnapshot workspaceConfig = null;
 
 	/***/
 	protected final List<IResourceTaskListener> listeners = new CopyOnWriteArrayList<>();
 
+	/*
+	 * Review feedback:
+	 *
+	 * Is this only ever used for the highlighting?
```
[the remark this answers is not in the document]
Comment (szarnekow) [reply plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/ResourceTaskManager.java:101]: Ok.

Reworded comment to
> Rethink the terminology: refresh is not something that is usually associated with an Xtext resource.

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/ResourceTaskManager.java [hunk 18]
```diff
@@ -233,6 +250,9 @@ public synchronized void changeSourceTextOfExistingContext(URI uri, int version,
 		Object queueId = getQueueIdForContext(rtc.getURI(), rtc.isTemporary());
 		return queuedExecutorService.submit(queueId, description, ci -> {
 			try {
+				if (!rtc.isAlive()) {
+					throw new CancellationException();
+				}
```
[the remark this answers is not in the document]
Comment (szarnekow) [reply plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/ResourceTaskManager.java:255]: I think the overhead for the isCancelled check on the rtc is not worth it. The isAlive is only for the very short window of opportunity when a task is closing a resource task context and another task was already submitted and captured the to-be-discarded context.

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/TextDocumentFrontend.java [hunk 19]
```diff
@@ -230,7 +222,8 @@ public void didClose(DidCloseTextDocumentParams params) {
 		}
 		XtextResource res = rtc.getResource();
 		XDocument doc = rtc.getDocument();
-		return Either.forLeft(documentSymbolService.getDefinitions(doc, res, params, resourceAccess, cancelIndicator));
+		return Either
+				.forLeft(documentSymbolService.getDefinitions(doc, res, params, resourceTaskManager, cancelIndicator));
```
[the remark this answers is not in the document]
Comment (szarnekow) [reply plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/TextDocumentFrontend.java:226]: :+1:

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/BuilderFrontend.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (szarnekow) [reply plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/BuilderFrontend.java:123]: Done

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/XURIBasedFileSystemAccessFactory.java (hunk 4 above)
Comment (szarnekow) [reply plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/XURIBasedFileSystemAccessFactory.java:34]: I tried to figure out which of the assumptions about the encapsulation of the builder package hold true. Actually I would expect that I could make all types package visible if I introduce a local Guice module for that package. So far that was not possible (e.g. due to N4jscCompiler). I made the class public again.

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/xtext/workspace/ProjectConfigSnapshot.java (hunk 5 above)
Comment (szarnekow) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/xtext/workspace/ProjectConfigSnapshot.java:69]: Added canonical methods to each `*Snapshot`.

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/xtext/workspace/WorkspaceConfigSnapshot.java (hunk 6 above)
Comment (szarnekow) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/xtext/workspace/WorkspaceConfigSnapshot.java:74]: I think having the same project twice in a workspace will be a source of hard to track-down bugs. Therefore I hardened the constraints.

**Comment by mmews-n4** [comment]
Thanks a lot! Builds are green enough. Merging.
